### PR TITLE
Fix circular dependency between modules

### PR DESCRIPTION
**Pull request by norihiro-w**
fix circular dependency between 
- MathLib and AssemblerLib
- AssemblerLib and NumLib
- NumLib and ProcessLib
- MeshLib and InsituLib

see current dependency graph https://doxygen.opengeosys.org/dir_6ccf91d5adff89be0d7026e347471e07.html

To fix them, I did the followings and removed AssemblerLib and InsituLib:
- merge AssemblerLib to NumLib (DOF and Assembler dirs)
- move DOF related stuff in MathLib/LinAlg to NumLib/DOF
- move NumericsConfig.h and GlobalSetup.h in ProcessLib to NumLib
- move InsituLib to MeshLib/Vtk

### CODE REVIEW

**Comment by norihiro-w** [comment]
and this is the new dependency graph
module

File: Applications/DataExplorer/DataView/ElementTreeModel.cpp [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
 #include "GeoLib/AABB.h"
 
-#include "InSituLib/VtkMappedMeshSource.h"
+#include "MeshLib/vtk/VtkMappedMeshSource.h"
```
Comment (bilke) [inline Applications/DataExplorer/DataView/ElementTreeModel.cpp]: `vtk` should be upper-case: `Vtk`

**Comment by bilke** [comment]
👍 for InSituLib-move.

File: Applications/DataExplorer/DataView/ElementTreeModel.cpp (hunk 1 above)
Comment (chleh) [inline Applications/DataExplorer/DataView/ElementTreeModel.cpp]: Why not all uppercase `VTK`?

File: MathLib/LinAlg/SparsityPattern.h [hunk 2]
```diff
@@ -12,12 +12,11 @@
...
 #include <vector>
 
-#include "ProcessLib/NumericsConfig.h"
-
 namespace MathLib
 {
 /// A vector telling how many nonzeros there are in each global matrix row.
-using SparsityPattern = std::vector<GlobalIndexType>;
+template <typename IndexType>
+using SparsityPattern = std::vector<IndexType>;
```
Comment (chleh) [inline MathLib/LinAlg/SparsityPattern.h]: I'm not immediately 100% sure if that should be a templated type alias...

File: NumLib/NumericsConfig.h [hunk 3]
```diff
@@ -12,6 +12,8 @@
 
 #include <type_traits>
 
+#include "MathLib/LinAlg/SparsityPattern.h"
```
Comment (chleh) [inline NumLib/NumericsConfig.h]: Why not move it to MathLib alongside with GlobalSetup? The two files only reference MathLib stuff, or am I wrong?

File: ProcessLib/DirichletBc.h [hunk 4]
```diff
@@ -12,19 +12,15 @@
...
 
 /// A dirichlet boundary condition is represented by a list of global indices
 /// with corresponding values.
 template <typename IndexType>
-struct DirichletBc final
-{
-    std::vector<IndexType> global_ids;
-    std::vector<double> values;
-};
+using DirichletBc = NumLib::IndexValueVector<IndexType>;
```
Comment (chleh) [inline ProcessLib/DirichletBc.h]: Is it worth keeping the file only for one typedef?

File: NumLib/ODESolver/NonlinearSolver-impl.h [hunk 5]
```diff
@@ -14,11 +14,11 @@
 
 #include "BaseLib/ConfigTree.h"
 #include "MathLib/LinAlg/BLAS.h"
-#include "AssemblerLib/GlobalMatrixProviders.h"
+#include "MathLib/LinAlg/VectorNorms.h"
```
Comment (chleh) [inline NumLib/ODESolver/NonlinearSolver-impl.h]: Is that needed?

File: NumLib/IndexValueVector.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+#ifndef NUM_LIB_INDEXVALUEVECTOR_H
+#define NUM_LIB_INDEXVALUEVECTOR_H
+
+#include <vector>
+
+namespace NumLib
+{
+
+template <typename IndexType>
+struct IndexValueVector final
```
Comment (chleh) [inline NumLib/IndexValueVector.h]: I wouldn't call it vector, since it's two vectors, but I don't have a better name for it right now.

File: NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.h [hunk 7]
```diff
@@ -10,7 +10,7 @@
 #ifndef NUMLIB_LOCAL_LLSQ_EXTRAPOLATOR_H
 #define NUMLIB_LOCAL_LLSQ_EXTRAPOLATOR_H
 
-#include "AssemblerLib/GlobalMatrixProviders.h"
+#include "NumLib/DOF/GlobalMatrixProviders.h"
```
Comment (chleh) [inline NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator.h]: I'd leave it in `MathLib`, since it only copes with Mathlib stuff. Or is that impossible?
In particular I wouldn't put it into the DOF directory, because it merely provides matrices and vectors.

File: Applications/ApplicationsLib/LinearSolverLibrarySetup.h [hunk 8]
```diff
@@ -19,7 +19,7 @@
 /// The default implementation is empty providing polymorphic behaviour when
 /// using this class.
 
-#include "MathLib/LinAlg/GlobalMatrixProviders.h"
+#include "AssemblerLib/GlobalMatrixProviders.h"
```
Comment (chleh) [inline Applications/ApplicationsLib/LinearSolverLibrarySetup.h]: This issue can be fixed IMHO in a better way by consistently naming the member functions of the DOF table. Then the DOF related methods of the matrix and vector providers would only get integer arguments and not the whole DOF table.

**Comment by chleh** [comment]
I think there should be a different "fix" for the matrix/vector related issues. I don't like to have them in AssemblerLib. Then everything you do with matrices and vectors will pull in AssemblerLib. I already discussed that with @endJunction some weeks ago.
But I'm mostly OK with your other changes.

File: Applications/DataExplorer/DataView/ElementTreeModel.cpp (hunk 1 above)
Comment (norihiro-w) [inline Applications/DataExplorer/DataView/ElementTreeModel.cpp]: i followed MeshLib/IO/VtkIO

**Comment by norihiro-w** [comment]
@chleh There is no AssemblerLib anymore and they are moved to NumLib/DOF and NumLib/Assembler. 
- NumLib/DOF includes DOF stuff and mat/vec files depending on DOF (moved from MathLib/LinAlg)
- NumLib/Assembler includes Executor.

Some functions in mat/vec files which are independent of DOF can be extracted from the files and can be moved back to MathLib/LinAlg. However the refactoring will not be a part of this PR.

**Comment by chleh** [comment]
OK, I can do the moving back to MathLib later on. I think the mat/vec code really belongs there.

**Comment by norihiro-w** [comment]
I will rework on this once Christoph's branch is merged, but comments are always welcome

File: NumLib/NumericsConfig.h (hunk 3 above)
Comment (norihiro-w) [inline NumLib/NumericsConfig.h]: Basically I would like to keep a module as general as possible. It's hard to do it for NumLib but MathLib is still possible with this way. My understanding is that these configuration are specifying how we use or specialize MathLib for our purpose, i.e. ogs6.

File: ProcessLib/DirichletBc.h (hunk 4 above)
Comment (norihiro-w) [inline ProcessLib/DirichletBc.h]: I personally don't care it. If you don't like it, may I ask you to do it in another PR?

File: NumLib/IndexValueVector.h (hunk 6 above)
Comment (norihiro-w) [inline NumLib/IndexValueVector.h]: me neither

**Comment by norihiro-w** [comment]
rebased

File: NumLib/IndexValueVector.h (hunk 6 above)
Comment (endJunction) [inline NumLib/IndexValueVector.h]: Just an idea: `KnownSolutions` or `KnownSolutionsVector` -- plays well with `applyKnownSolution`.

File: NumLib/NumericsConfig.h (hunk 3 above)
Comment (chleh) [inline NumLib/NumericsConfig.h]: OK, let's discuss that after this PR.

**Comment by chleh** [comment]
:+1:

File: NumLib/IndexValueVector.h (hunk 6 above)
Comment (norihiro-w) [inline NumLib/IndexValueVector.h]: I think this data type can also be used for Neumann or other conditions. In that case, the values are not known solutions any more.
